stage0/src/acpi.rs

```diff
@@ -0,0 +1,411 @@
+    }
+}
+
+impl Debug for AddChecksum {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        f.debug_struct("AddChecksum")
+            .field("file", &self.file())
+            .field("offset", &self.offset)
+            .field("start", &self.start)
+            .field("length", &self.length)
+            .finish()
+    }
+}
+
+/// COMMAND_WRITE_POINTER - write the fw_cfg file (originating from `dest_file`) at `offset`, by
+/// adding a pointer to `src_offset` within the table originating from `src_file`.
+///
+/// 1,2,4 or 8 byte unsigned addition is used depending on `size`.
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct WritePointer {
+    dest_file: RomfileName,
+    src_file: RomfileName,
+    dst_offset: u32,
+    src_offset: u32,
+    size: u8,
+}
+static_assertions::assert_eq_size!(WritePointer, Pad);
+
+impl WritePointer {
+    pub fn dest_file(&self) -> &CStr {
+        CStr::from_bytes_until_nul(&self.dest_file).unwrap()
+    }
+
+    pub fn src_file(&self) -> &CStr {
+        CStr::from_bytes_until_nul(&self.src_file).unwrap()
+    }
+
+    fn invoke(&self) -> Result<(), &'static str> {
+        log::debug!("{:?}", self);
+        Err("COMMAND_WRITE_POINTER is not supported")
+    }
+}
+
+impl Debug for WritePointer {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        f.debug_struct("WritePointer")
+            .field("dest_file", &self.dest_file())
+            .field("src_file", &self.src_file())
+            .field("dst_offset", &self.dst_offset)
+            .field("src_offset", &self.src_offset)
+            .field("size", &self.size)
+            .finish()
+    }
+}
+
+type Pad = [u8; 124];
+
+#[repr(C)]
+union Body {
+    allocate: Allocate,
+    pointer: AddPointer,
+    checksum: AddChecksum,
+    wr_pointer: WritePointer,
+    padding: Pad,
+}
+
+impl Default for Body {
+    fn default() -> Self {
+        Body {
+            padding: [Default::default(); 124],
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Default)]
+struct RomfileCommand {
+    tag: u32,
+    body: Body,
+}
+
+#[derive(Debug)]
+enum Command<'a> {
+    Allocate(&'a Allocate),
+    AddPointer(&'a AddPointer),
+    AddChecksum(&'a AddChecksum),
+    WritePointer(&'a WritePointer),
+}
+
+impl Command<'_> {
+    pub fn invoke(&self, fwcfg: &mut FwCfg) -> Result<(), &'static str> {
+        match self {
+            Command::Allocate(allocate) => allocate.invoke(fwcfg),
+            Command::AddPointer(add_pointer) => add_pointer.invoke(),
+            Command::AddChecksum(add_checksum) => add_checksum.invoke(),
+            Command::WritePointer(write_pointer) => write_pointer.invoke(),
+        }
+    }
+}
+
+impl RomfileCommand {
+    fn tag(&self) -> Option<CommandTag> {
+        CommandTag::from_repr(self.tag)
+    }
+
+    fn extract(&self) -> Command<'_> {
+        // Safety: we extract the value out of the union based on the tag value, which is safe to
+        // do.
+        match self.tag().unwrap() {
```

Comment:
Is crashing OK here? Since `extract` is only used in a `invoke` which returns a result, perhaps it is better for `extract` to also return a result and use `.ok_or(...)?` insead of `unwrap`.

Reply:
I'd argue it is, because this means we got something we don't support from the `etc/table-loader`. However, I agree that it'd be more consistent to push the error handling up to one place if we're doing it everywhere anyway.